### PR TITLE
Fix manageemnt failed data sent

### PR DESCRIPTION
**Pull request by smurf-U**
Losing data whenever an API call fails or there is any error from Airbrake. Because Pybrake create a local variable, copy the data, and set the class variable to None, when PyBrake sends statistics to AirBrake. However, when Pybrake encounter API failure or errors, Pybrake do not update or reassign statistics data to the class verifiable.

To fix this, creating a circular buffer or invalidate list type variable and appending json data to it while the API call fails, then at flush time sending current stats first and then checking that the list is not empty, pop data and sending those data as well if it fails, putting that again in it.

### CODE REVIEW

File: src/pybrake/notifier.py [hunk 1]
```diff
@@ -73,13 +72,19 @@ def __init__(
                 default value [re.compile("password"), re.compile("secret")]
         :param remote_config: Set up configuration from the Airbrake server,
                 default value False
+        :param max_failed_queue: Set up length of failed stats queue
```
Comment (kyrylo) [inline src/pybrake/notifier.py:75]: What do you think if we renamed this to `max_backlog_size`? Technically, unset metrics is a backlog waiting to be sent.

File: src/pybrake/notifier.py [hunk 2]
```diff
@@ -73,13 +72,19 @@ def __init__(
...
+        :param manage_failed_stats: If manage_failed_stats set as true then
```
Comment (kyrylo) [inline src/pybrake/notifier.py:76]: I would call this option `backlog`. When I read:

```
* max_failed_queue
* manage_failed_stats
```

...I don't see a connection between these two options.

If I read:

```
* max_backlog_size
* backlog
```

...It's obvious there's some connection.


`backlog` can also be named `backlog_enabled`

What do you think?

File: src/pybrake/notifier.py [hunk 3]
```diff
@@ -116,6 +121,15 @@ def __init__(
         self._context = _CONTEXT.copy()
         self._context["rootDirectory"] = kwargs.get("root_directory",
                                                     os.getcwd())
+        self._failed_stats = None
+        if self.config.get('manage_failed_stats'):
+            self._failed_stats = FailedStats(
```
Comment (kyrylo) [inline src/pybrake/notifier.py:126]: `FailedStats` could be named `Backlog` if you agree with the option renaming.

File: src/pybrake/notifier.py [hunk 4]
```diff
@@ -116,6 +121,15 @@ def __init__(
...
+                interval=FLUSH_PERIOD,
+                header=self._ab_headers,
+                url=self._ab_url,
+                method="POST",
```
Comment (kyrylo) [inline src/pybrake/notifier.py:130]: Is it important to make `method` configurable?

File: src/pybrake/notifier.py [hunk 5]
```diff
@@ -228,9 +242,13 @@ def _send_notice_sync(self, notice):
             resp = urllib.request.urlopen(req, timeout=5)
         except urllib.error.HTTPError as err:
             resp = err
+            if self.config.get('manage_failed_stats'):
```
Comment (kyrylo) [inline src/pybrake/notifier.py:245]:
```suggestion
            if self._failed_stats is not None:
```

File: src/pybrake/failed_stats.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
+import time
+from collections import deque
+from threading import Timer
+
+from . import metrics
+
+
+class FailedStats(Timer):
```
Comment (kyrylo) [inline src/pybrake/failed_stats.py:8]: It would be nice to add tests.

Comment (kyrylo) [reply src/pybrake/failed_stats.py:8]: We should probably call `cancel()` on a timer.

Alternatively, you can inherit from `Thread` since you use a while loop https://stackoverflow.com/a/9812806/511205

I'm not a Python expert, so it's up to you.

File: src/pybrake/metrics.py [hunk 7]
```diff
@@ -136,3 +137,57 @@ def _resume(self):
         if not self._paused():
             return
         self.start_time = pytime.time()
+
+
+def send(url, headers, failed_stats, payload=None, method=None):
```
Comment (kyrylo) [inline src/pybrake/metrics.py:142]: Can you please perform this refactoring in a separate PR? We will merge it prior to merging this one. This looks unrelated to adding the backlog aka failed stats.

File: src/pybrake/notifier.py (hunk 4 above)
Comment (smurf-U) [reply src/pybrake/notifier.py:130]: Method used in the `send` function, It has already been configured. However, because this API is a POST call, I passed hard coded over here.

File: src/pybrake/failed_stats.py (hunk 6 above)
Comment (smurf-U) [reply src/pybrake/failed_stats.py]: I'm working on test cases. Once all adjustments were approved, I thought an additional PR would be raised for that.

File: src/pybrake/metrics.py (hunk 7 above)
Comment (smurf-U) [reply src/pybrake/metrics.py]: Sure

**Comment by smurf-U** [comment]
PTAL

**Review by kyrylo — APPROVED** [review]
